src/peripherals-provider.ts

```diff
@@ -0,0 +1,93 @@
+import * as vscode from 'vscode';
+import * as manifest from './manifest';
+import { SvdResolver } from './svd-resolver';
+import { PeripheralNode, PeripheralOptions } from './views/nodes/peripheralnode';
+import { SvdData, SVDParser } from './svd-parser';
+import { parseStringPromise } from 'xml2js';
+import { readFromUrl } from './utils';
+
+const pathToUri = (path: string): vscode.Uri => {
+    try {
+        return vscode.Uri.file(path);
+    } catch (e) {
+        return vscode.Uri.parse(path);
+    }
+};
+
+export class PeripheralsProvider {
+    constructor(protected svdResolver: SvdResolver) {
+    }
+
+    public async getPeripherals(session: vscode.DebugSession, wsFolderPath?: vscode.Uri): Promise<PeripheralNode[] | undefined> {
+        const getPeripheralsConfig = vscode.workspace.getConfiguration(manifest.PACKAGE_NAME).get<string>(manifest.CONFIG_GET_PERIPHERALS) || manifest.DEFAULT_GET_PERIPHERALS;
+        const getPeripheralsCommand = session.configuration[getPeripheralsConfig];
+
+        let thresh = session.configuration[manifest.CONFIG_ADDRGAP];
+        if (!thresh) {
+            thresh = vscode.workspace.getConfiguration(manifest.PACKAGE_NAME).get<number>(manifest.CONFIG_ADDRGAP) || manifest.DEFAULT_ADDRGAP;
+        }
+
+        if (((typeof thresh) === 'number') && (thresh < 0)) {
+            thresh = -1;     // Never merge register reads even if adjacent
+        } else {
+            // Set the threshold between 0 and 32, with a default of 16 and a mukltiple of 8
+            thresh = ((((typeof thresh) === 'number') ? Math.max(0, Math.min(thresh, 32)) : 16) + 7) & ~0x7;
```

Comment:
~BTW I had to read this line a couple of times to grok it. I think four opening brackets in an expression is part of the problem, the other part that is that `+ 7) & ~0x7` part of the expression applies even when it is a constant `16`.~ I see now this is refactored code, not new code.